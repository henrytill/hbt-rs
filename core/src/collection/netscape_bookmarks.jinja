<!DOCTYPE NETSCAPE-Bookmark-file-1>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
<TITLE>Bookmarks</TITLE>
<H1>Bookmarks</H1>
<DL><p>
{%- for entity in entities %}
{%- set title = entity.names | first | default(entity.uri) %}
{%- set tags_str = entity.labels | join(",") if entity.labels else none %}
{%- set last_modified = entity.updatedAt | first if entity.updatedAt else none %}
    <DT><A HREF="{{ entity.uri }}" ADD_DATE="{{ entity.createdAt }}"{% if last_modified %} LAST_MODIFIED="{{ last_modified }}"{% endif %}{% if tags_str %} TAGS="{{ tags_str }}"{% endif %}{% if not entity.shared %} PRIVATE="1"{% endif %}{% if entity.toRead %} TOREAD="1"{% endif %}{% if entity.isFeed %} FEED="true"{% endif %}{% if entity.lastVisitedAt %} LAST_VISIT="{{ entity.lastVisitedAt }}"{% endif %}>{{ title }}</A>
{%- if entity.extended %}
    <DD>{{ entity.extended }}
{%- endif %}
{%- endfor %}
</DL><p>
