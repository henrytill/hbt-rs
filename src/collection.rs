use std::{
    collections::{HashMap, HashSet},
    hash::{Hash, Hasher},
    ops::{Index, IndexMut},
};

use serde::Serialize;
use time::Date;
use url::Url;

/// An [`Id`] is a unique identifier for an [`Entity`].
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Id(usize);

impl Id {
    const fn new(id: usize) -> Id {
        Id(id)
    }
}

impl From<Id> for usize {
    fn from(id: Id) -> usize {
        id.0
    }
}

/// A [`Name`] describes an [`Entity`].
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Name(String);

impl Name {
    pub const fn new(name: String) -> Name {
        Name(name)
    }

    pub fn as_str(&self) -> &str {
        self.0.as_str()
    }
}

impl Hash for Name {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

impl From<String> for Name {
    fn from(name: String) -> Name {
        Name(name)
    }
}

#[cfg(test)]
impl From<&str> for Name {
    fn from(name: &str) -> Name {
        Name(name.into())
    }
}

/// A [`Label`] is a label that can be attached to an [`Entity`].
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Label(String);

impl Label {
    pub const fn new(name: String) -> Label {
        Label(name)
    }

    pub fn as_str(&self) -> &str {
        self.0.as_str()
    }
}

impl Hash for Label {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

impl From<String> for Label {
    fn from(name: String) -> Label {
        Label(name)
    }
}

#[cfg(test)]
impl From<&str> for Label {
    fn from(name: &str) -> Label {
        Label(name.into())
    }
}

/// An [`Entity`] is a page in the collection.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Entity {
    url: Url,
    created_at: Date,
    updated_at: Vec<Date>,
    names: HashSet<Name>,
    labels: HashSet<Label>,
}

impl Entity {
    pub fn new(
        url: Url,
        created_at: Date,
        maybe_name: Option<Name>,
        labels: HashSet<Label>,
    ) -> Entity {
        let updated_at = Vec::new();
        let names = maybe_name.into_iter().collect();
        Entity { url, created_at, updated_at, names, labels }
    }

    pub fn update(
        &mut self,
        updated_at: Date,
        names: HashSet<Name>,
        labels: HashSet<Label>,
    ) -> &mut Entity {
        if updated_at < self.created_at {
            self.updated_at.push(self.created_at);
            self.created_at = updated_at;
        } else {
            self.updated_at.push(updated_at);
        }
        self.names.extend(names);
        self.labels.extend(labels);
        self
    }

    pub fn merge(&mut self, other: Entity) -> &mut Entity {
        self.update(other.created_at, other.names, other.labels)
    }

    pub fn url(&self) -> &Url {
        &self.url
    }

    pub fn created_at(&self) -> &Date {
        &self.created_at
    }

    pub fn updated_at(&self) -> &[Date] {
        &self.updated_at
    }

    pub fn names(&self) -> &HashSet<Name> {
        &self.names
    }

    pub fn labels(&self) -> &HashSet<Label> {
        &self.labels
    }
}

pub type Edges = Vec<Id>;

/// A collection of entities.
///
/// This is a graph structure where a nodes are represented by a vector of entities and edges are
/// represented by an adjacency list.
#[derive(Debug)]
pub struct Collection {
    nodes: Vec<Entity>,
    edges: Vec<Edges>,
    urls: HashMap<Url, Id>,
}

impl Index<Id> for Vec<Entity> {
    type Output = Entity;

    fn index(&self, id: Id) -> &Self::Output {
        &self[id.0]
    }
}

impl IndexMut<Id> for Vec<Entity> {
    fn index_mut(&mut self, id: Id) -> &mut Self::Output {
        &mut self[id.0]
    }
}

impl Index<Id> for Vec<Edges> {
    type Output = Edges;

    fn index(&self, id: Id) -> &Self::Output {
        &self[id.0]
    }
}

impl IndexMut<Id> for Vec<Edges> {
    fn index_mut(&mut self, id: Id) -> &mut Self::Output {
        &mut self[id.0]
    }
}

impl Collection {
    pub fn new() -> Collection {
        let nodes = Vec::new();
        let edges = Vec::new();
        let urls = HashMap::new();
        Collection { nodes, edges, urls }
    }

    pub fn len(&self) -> usize {
        let len = self.nodes.len();
        assert_eq!(len, self.edges.len());
        len
    }

    pub fn is_empty(&self) -> bool {
        let is_empty = self.nodes.is_empty();
        assert_eq!(is_empty, self.edges.is_empty());
        is_empty
    }

    pub fn contains(&self, url: &Url) -> bool {
        self.urls.contains_key(url)
    }

    pub fn id(&self, url: &Url) -> Option<Id> {
        self.urls.get(url).copied()
    }

    pub fn insert(&mut self, entity: Entity) -> Id {
        let id = Id::new(self.len());
        self.nodes.push(entity);
        self.edges.push(Vec::new());
        let url = self.nodes[id].url().to_owned();
        self.urls.insert(url, id);
        id
    }

    pub fn upsert(&mut self, other: Entity) -> Id {
        if let Some(id) = self.id(other.url()) {
            let entity = &mut self.nodes[id];
            entity.merge(other);
            id
        } else {
            self.insert(other)
        }
    }

    pub fn add_edge(&mut self, from: Id, to: Id) {
        let from_edges = &mut self.edges[from];
        if from_edges.contains(&to) {
            return;
        }
        from_edges.push(to);
    }

    pub fn add_edges(&mut self, from: Id, to: Id) {
        self.add_edge(from, to);
        self.add_edge(to, from)
    }

    pub fn entity(&self, id: Id) -> &Entity {
        &self.nodes[id]
    }

    pub fn entity_mut(&mut self, id: Id) -> &mut Entity {
        &mut self.nodes[id]
    }

    pub fn edges(&self, id: Id) -> &[Id] {
        &self.edges[id]
    }

    pub fn entities(&self) -> &[Entity] {
        &self.nodes
    }
}

impl Default for Collection {
    fn default() -> Collection {
        Collection::new()
    }
}
