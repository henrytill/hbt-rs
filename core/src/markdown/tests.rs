use std::vec;

use time::macros::datetime;

use super::*;
use crate::collection::Time;

const TEST_EMPTY: &str = "";

#[test]
fn test_empty() {
    let collection = parse(TEST_EMPTY).unwrap();
    assert!(collection.is_empty());
}

const TEST_NO_DATE: &str = "\
- [Foo](https://foo.com)
";

#[test]
fn test_no_date() {
    let expected = "missing date";
    let actual = parse(TEST_NO_DATE).unwrap_err().to_string();
    assert_eq!(expected, actual);
}

const TEST_ONLY_DATE: &str = "\
# November 15, 2023
";

#[test]
fn test_only_date() {
    let collection = parse(TEST_ONLY_DATE).unwrap();
    assert!(collection.is_empty());
}

const TEST_NO_LABELS: &str = "\
# November 15, 2023

- [Foo](https://foo.com)
- [Bar](https://bar.com)
";

#[test]
fn test_no_labels() {
    let collection = parse(TEST_NO_LABELS).unwrap();
    assert_eq!(collection.len(), 2);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Some(Name::from("Bar")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
    }
}

const TEST_NO_URL: &str = "\
# November 15, 2023

- Foo
";

#[test]
fn test_no_url() {
    let collection = parse(TEST_NO_URL).unwrap();
    assert!(collection.is_empty());
}

const TEST_NO_TITLE: &str = "\
# November 15, 2023

- <https://foo.com>
";

#[test]
fn test_no_title() {
    let collection = parse(TEST_NO_TITLE).unwrap();

    assert_eq!(collection.len(), 1);

    let expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        datetime!(2023-11-15 0:00 UTC).into(),
        Default::default(),
        Default::default(),
    );
    let id = collection.id(expected.url()).unwrap();
    assert_eq!(&expected, collection.entity(id));
}

const TEST_INDENTED: &str = "\
# November 15, 2023

  - [Foo](https://foo.com)
";

#[test]
fn test_indented() {
    let collection = parse(TEST_INDENTED).unwrap();

    assert_eq!(collection.len(), 1);

    let expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        datetime!(2023-11-15 0:00 UTC).into(),
        Some(Name::from("Foo")),
        Default::default(),
    );
    let id = collection.id(expected.url()).unwrap();
    assert_eq!(&expected, collection.entity(id));
}

const TEST_INDENTED_DOUBLE: &str = "\
# November 15, 2023

    - [Foo](https://foo.com)
";

#[test]
fn test_indented_double() {
    let collection = parse(TEST_INDENTED_DOUBLE).unwrap();
    assert!(collection.is_empty());
}

const TEST_PARENT: &str = "\
# November 15, 2023

- [Foo](https://foo.com)
  - [Bar](https://bar.com)
";

#[test]
fn test_parent() {
    let collection = parse(TEST_PARENT).unwrap();
    assert_eq!(collection.len(), 2);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    let foo_expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        expected_time,
        Some(Name::from("Foo")),
        Default::default(),
    );
    let foo_id = collection.id(foo_expected.url()).unwrap();
    assert_eq!(&foo_expected, collection.entity(foo_id));
    let foo_edges = collection.edges(foo_id);
    assert_eq!(foo_edges.len(), 1);

    let bar_expected = Entity::new(
        Url::parse("https://bar.com").unwrap(),
        expected_time,
        Some(Name::from("Bar")),
        Default::default(),
    );
    let bar_id = collection.id(bar_expected.url()).unwrap();
    assert_eq!(&bar_expected, collection.entity(bar_id));
    let bar_edges = collection.edges(bar_id);
    assert_eq!(bar_edges.len(), 1);

    assert_eq!(foo_edges, vec![bar_id]);
    assert_eq!(bar_edges, vec![foo_id]);
}

const TEST_PARENTS: &str = "\
# November 15, 2023

- [Foo](https://foo.com)
  - [Bar](https://bar.com)
    - [Baz](https://baz.com)
";

#[test]
fn test_parents() {
    let collection = parse(TEST_PARENTS).unwrap();
    assert_eq!(collection.len(), 3);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    let foo_expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        expected_time,
        Some(Name::from("Foo")),
        Default::default(),
    );
    let foo_id = collection.id(foo_expected.url()).unwrap();
    assert_eq!(&foo_expected, collection.entity(foo_id));
    let foo_edges = collection.edges(foo_id);
    assert_eq!(foo_edges.len(), 1);

    let bar_expected = Entity::new(
        Url::parse("https://bar.com").unwrap(),
        expected_time,
        Some(Name::from("Bar")),
        Default::default(),
    );
    let bar_id = collection.id(bar_expected.url()).unwrap();
    assert_eq!(&bar_expected, collection.entity(bar_id));
    let bar_edges = collection.edges(bar_id);
    assert_eq!(bar_edges.len(), 2);

    let baz_expected = Entity::new(
        Url::parse("https://baz.com").unwrap(),
        expected_time,
        Some(Name::from("Baz")),
        Default::default(),
    );
    let baz_id = collection.id(baz_expected.url()).unwrap();
    assert_eq!(&baz_expected, collection.entity(baz_id));
    let baz_edges = collection.edges(baz_id);
    assert_eq!(baz_edges.len(), 1);

    assert_eq!(foo_edges, vec![bar_id]);
    assert_eq!(bar_edges, vec![foo_id, baz_id]);
    assert_eq!(baz_edges, vec![bar_id]);
}

const TEST_PARENTS_INDENTED: &str = "\
# November 15, 2023

  - [Foo](https://foo.com)
    - [Bar](https://bar.com)
      - [Baz](https://baz.com)
";

#[test]
fn test_parents_indented() {
    let collection = parse(TEST_PARENTS_INDENTED).unwrap();
    assert_eq!(collection.len(), 3);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    let foo_expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        expected_time,
        Some(Name::from("Foo")),
        Default::default(),
    );
    let foo_id = collection.id(foo_expected.url()).unwrap();
    assert_eq!(&foo_expected, collection.entity(foo_id));
    let foo_edges = collection.edges(foo_id);
    assert_eq!(foo_edges.len(), 1);

    let bar_expected = Entity::new(
        Url::parse("https://bar.com").unwrap(),
        expected_time,
        Some(Name::from("Bar")),
        Default::default(),
    );
    let bar_id = collection.id(bar_expected.url()).unwrap();
    assert_eq!(&bar_expected, collection.entity(bar_id));
    let bar_edges = collection.edges(bar_id);
    assert_eq!(bar_edges.len(), 2);

    let baz_expected = Entity::new(
        Url::parse("https://baz.com").unwrap(),
        expected_time,
        Some(Name::from("Baz")),
        Default::default(),
    );
    let baz_id = collection.id(baz_expected.url()).unwrap();
    assert_eq!(&baz_expected, collection.entity(baz_id));
    let baz_edges = collection.edges(baz_id);
    assert_eq!(baz_edges.len(), 1);

    assert_eq!(foo_edges, vec![bar_id]);
    assert_eq!(bar_edges, vec![foo_id, baz_id]);
    assert_eq!(baz_edges, vec![bar_id]);
}

const TEST_SINGLE_PARENT: &str = "\
# November 15, 2023

- [Foo](https://foo.com)
  - [Bar](https://bar.com)
  - [Baz](https://baz.com)
  - [Quux](https://quux.com)
";

#[test]
fn test_single_parent() {
    let collection = parse(TEST_SINGLE_PARENT).unwrap();
    assert_eq!(collection.len(), 4);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    let foo_expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        expected_time,
        Some(Name::from("Foo")),
        Default::default(),
    );
    let foo_id = collection.id(foo_expected.url()).unwrap();
    assert_eq!(&foo_expected, collection.entity(foo_id));
    let foo_edges = collection.edges(foo_id);
    assert_eq!(foo_edges.len(), 3);

    let bar_expected = Entity::new(
        Url::parse("https://bar.com").unwrap(),
        expected_time,
        Some(Name::from("Bar")),
        Default::default(),
    );

    let bar_id = collection.id(bar_expected.url()).unwrap();
    assert_eq!(&bar_expected, collection.entity(bar_id));
    let bar_edges = collection.edges(bar_id);
    assert_eq!(bar_edges.len(), 1);

    let baz_expected = Entity::new(
        Url::parse("https://baz.com").unwrap(),
        expected_time,
        Some(Name::from("Baz")),
        Default::default(),
    );
    let baz_id = collection.id(baz_expected.url()).unwrap();
    assert_eq!(&baz_expected, collection.entity(baz_id));
    let baz_edges = collection.edges(baz_id);
    assert_eq!(baz_edges.len(), 1);

    let quux_expected = Entity::new(
        Url::parse("https://quux.com").unwrap(),
        expected_time,
        Some(Name::from("Quux")),
        Default::default(),
    );
    let quux_id = collection.id(quux_expected.url()).unwrap();
    assert_eq!(&quux_expected, collection.entity(quux_id));
    let quux_edges = collection.edges(quux_id);
    assert_eq!(quux_edges.len(), 1);

    assert_eq!(foo_edges, vec![bar_id, baz_id, quux_id]);
    assert_eq!(bar_edges, vec![foo_id]);
    assert_eq!(baz_edges, vec![foo_id]);
    assert_eq!(quux_edges, vec![foo_id]);
}

const TEST_INVERTED_PARENT: &str = "\
# November 15, 2023

  - [Foo](https://foo.com)
- [Bar](https://bar.com)
";

#[test]
fn test_no_parent() {
    let collection = parse(TEST_INVERTED_PARENT).unwrap();
    assert_eq!(collection.len(), 2);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Some(Name::from("Bar")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }
}

const TEST_INVERTED_SINGLE_PARENT: &str = "\
# November 15, 2023

  - [Foo](https://foo.com)
  - [Bar](https://bar.com)
- [Baz](https://baz.com)
";

#[test]
fn test_inverted_parents() {
    let collection = parse(TEST_INVERTED_SINGLE_PARENT).unwrap();
    assert_eq!(collection.len(), 3);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Some(Name::from("Bar")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://baz.com").unwrap(),
            expected_time,
            Some(Name::from("Baz")),
            Default::default(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }
}

const TEST_LABEL: &str = "\
# November 15, 2023

## Foo

- [Foo](https://foo.com)
- [Bar](https://bar.com)
";

#[test]
fn test_label() {
    let collection = parse(TEST_LABEL).unwrap();
    assert_eq!(collection.len(), 2);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();
    let expected_labels = [Label::from("Foo")];

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            expected_labels.iter().cloned().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Some(Name::from("Bar")),
            expected_labels.into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }
}

const TEST_LABELS: &str = "\
# November 15, 2023

## Foo

- [Foo](https://foo.com)
- [Bar](https://bar.com)

## Baz

- [Baz](https://baz.com)
- [Quux](https://quux.com)
";

#[test]
fn test_labels() {
    let collection = parse(TEST_LABELS).unwrap();
    assert_eq!(collection.len(), 4);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();
    let expected_labels = [Label::from("Foo")];

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            expected_labels.iter().cloned().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Some(Name::from("Bar")),
            expected_labels.iter().cloned().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    let expected_labels = [Label::from("Baz")];

    {
        let expected = Entity::new(
            Url::parse("https://baz.com").unwrap(),
            expected_time,
            Some(Name::from("Baz")),
            expected_labels.iter().cloned().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://quux.com").unwrap(),
            expected_time,
            Some(Name::from("Quux")),
            expected_labels.into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }
}

const TEST_MULTIPLE_LABELS: &str = "\
# November 15, 2023

## Foo

- [Foo](https://foo.com)

### Bar

- [Bar](https://bar.com)

#### Baz

- [Baz](https://baz.com)
";

#[test]
fn test_multiple_labels() {
    let collection = parse(TEST_MULTIPLE_LABELS).unwrap();
    assert_eq!(collection.len(), 3);

    let expected_time = datetime!(2023-11-15 0:00 UTC).into();

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            [Label::from("Foo")].into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Some(Name::from("Bar")),
            [Label::from("Foo"), Label::from("Bar")].into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://baz.com").unwrap(),
            expected_time,
            Some(Name::from("Baz")),
            [Label::from("Foo"), Label::from("Bar"), Label::from("Baz")].into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }
}

const TEST_UPDATE: &str = "\
# December 5, 2023

## Foo

- [Foo](https://foo.com)

# December 6, 2023

## Bar

- [Bar](https://foo.com)
";

#[test]
fn test_update() {
    let collection = parse(TEST_UPDATE).unwrap();
    assert_eq!(collection.len(), 1);

    {
        let mut expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            datetime!(2023-12-5 0:00 UTC).into(),
            Some(Name::from("Foo")),
            [Label::from("Foo")].into_iter().collect(),
        );

        expected.update(
            datetime!(2023-12-6 0:00 UTC).into(),
            [Name::from("Bar")].into_iter().collect(),
            [Label::from("Bar")].into_iter().collect(),
        );

        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
    }
}

const TEST_DESCENDING_DATES: &str = "\
# December 6, 2023

## Foo

- [Foo](https://foo.com)

# December 5, 2023

## Bar

- [Bar](https://foo.com)
";

#[test]
fn test_descending_dates() {
    let collection = parse(TEST_DESCENDING_DATES).unwrap();
    assert_eq!(collection.len(), 1);

    let mut expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        datetime!(2023-12-6 0:00 UTC).into(),
        Some(Name::from("Foo")),
        [Label::from("Foo")].into_iter().collect(),
    );

    expected.update(
        datetime!(2023-12-5 0:00 UTC).into(),
        [Name::from("Bar")].into_iter().collect(),
        [Label::from("Bar")].into_iter().collect(),
    );

    let id = collection.id(expected.url()).unwrap();
    let actual = collection.entity(id);
    assert_eq!(&expected, actual);
    assert_eq!(actual.created_at(), &Time::new(datetime!(2023-12-5 0:00 UTC)));
    assert_eq!(actual.updated_at(), &[Time::new(datetime!(2023-12-6 0:00 UTC))]);
}

const TEST_MIXED_DATES: &str = "\
# December 6, 2023

## Foo

- [Foo](https://foo.com)

# December 5, 2023

## Bar

- [Bar](https://foo.com)

# December 7, 2023

## Baz

- [Baz](https://foo.com)
";

#[test]
fn test_mixed_dates() {
    let collection = parse(TEST_MIXED_DATES).unwrap();
    assert_eq!(collection.len(), 1);

    let mut expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        datetime!(2023-12-6 0:00 UTC).into(),
        Some(Name::from("Foo")),
        [Label::from("Foo")].into_iter().collect(),
    );

    expected.update(
        datetime!(2023-12-5 0:00 UTC).into(),
        [Name::from("Bar")].into_iter().collect(),
        [Label::from("Bar")].into_iter().collect(),
    );

    expected.update(
        datetime!(2023-12-7 0:00 UTC).into(),
        [Name::from("Baz")].into_iter().collect(),
        [Label::from("Baz")].into_iter().collect(),
    );

    let id = collection.id(expected.url()).unwrap();
    let actual = collection.entity(id);
    assert_eq!(&expected, actual);
    assert_eq!(actual.created_at(), &Time::new(datetime!(2023-12-5 0:00 UTC)));
    assert_eq!(
        actual.updated_at(),
        &[Time::new(datetime!(2023-12-6 0:00 UTC)), Time::new(datetime!(2023-12-7 0:00 UTC))]
    );
}

// Original tests below

const TEST_BASIC: &str = "\
# November 16, 2023

## Foo

- [Foo](https://foo.com)

### Bar

- <https://bar.com>

## Misc

- [Hello, world!](https://example.com/)
";

#[test]
fn test_basic() {
    let collection = parse(TEST_BASIC).unwrap();
    assert_eq!(collection.len(), 3);

    let expected_time = datetime!(2023-11-16 0:00 UTC).into();

    {
        let expected = Entity::new(
            Url::parse("https://foo.com").unwrap(),
            expected_time,
            Some(Name::from("Foo")),
            [Label::from("Foo")].into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://bar.com").unwrap(),
            expected_time,
            Default::default(),
            [Label::from("Foo"), Label::from("Bar")].into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }

    {
        let expected = Entity::new(
            Url::parse("https://example.com").unwrap(),
            expected_time,
            Some(Name::from("Hello, world!")),
            [Label::from("Misc")].into_iter().collect(),
        );
        let id = collection.id(expected.url()).unwrap();
        assert_eq!(&expected, collection.entity(id));
        assert!(collection.edges(id).is_empty());
    }
}

const TEST_NESTED: &str = "\
# November 17, 2023

## Foo

- [Foo](https://foo.com)
  - <https://bar.com>
  - [Hello, world!](https://example.com/)
    - [Quux](https://quux.com)
  - <https://baz.com>
";

#[test]
fn test_nested() {
    let collection = parse(TEST_NESTED).unwrap();
    assert_eq!(collection.len(), 5);

    let expected_time = datetime!(2023-11-17 0:00 UTC).into();
    let expected_labels = [Label::from("Foo")];

    let foo_expected = Entity::new(
        Url::parse("https://foo.com").unwrap(),
        expected_time,
        Some(Name::from("Foo")),
        expected_labels.iter().cloned().collect(),
    );
    let foo_id = collection.id(foo_expected.url()).unwrap();
    assert_eq!(&foo_expected, collection.entity(foo_id));
    let foo_edges = collection.edges(foo_id);
    assert_eq!(foo_edges.len(), 3);

    let bar_expected = Entity::new(
        Url::parse("https://bar.com").unwrap(),
        expected_time,
        Default::default(),
        expected_labels.iter().cloned().collect(),
    );
    let bar_id = collection.id(bar_expected.url()).unwrap();
    assert_eq!(&bar_expected, collection.entity(bar_id));
    let bar_edges = collection.edges(bar_id);
    assert_eq!(bar_edges.len(), 1);

    let hello_expected = Entity::new(
        Url::parse("https://example.com").unwrap(),
        expected_time,
        Some(Name::from("Hello, world!")),
        expected_labels.iter().cloned().collect(),
    );
    let hello_id = collection.id(hello_expected.url()).unwrap();
    assert_eq!(&hello_expected, collection.entity(hello_id));
    let hello_edges = collection.edges(hello_id);
    assert_eq!(hello_edges.len(), 2);

    let quux_expected = Entity::new(
        Url::parse("https://quux.com").unwrap(),
        expected_time,
        Some(Name::from("Quux")),
        expected_labels.iter().cloned().collect(),
    );
    let quux_id = collection.id(quux_expected.url()).unwrap();
    assert_eq!(&quux_expected, collection.entity(quux_id));
    let quux_edges = collection.edges(quux_id);
    assert_eq!(quux_edges.len(), 1);

    let baz_expected = Entity::new(
        Url::parse("https://baz.com").unwrap(),
        expected_time,
        Default::default(),
        expected_labels.into_iter().collect(),
    );
    let baz_id = collection.id(baz_expected.url()).unwrap();
    assert_eq!(&baz_expected, collection.entity(baz_id));
    let baz_edges = collection.edges(baz_id);
    assert_eq!(baz_edges.len(), 1);

    assert_eq!(foo_edges, vec![bar_id, hello_id, baz_id]);
    assert_eq!(bar_edges, vec![foo_id]);
    assert_eq!(hello_edges, vec![foo_id, quux_id]);
    assert_eq!(quux_edges, vec![hello_id]);
    assert_eq!(baz_edges, vec![foo_id]);
}
