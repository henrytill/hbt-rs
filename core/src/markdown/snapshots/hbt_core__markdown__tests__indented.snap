---
source: core/src/markdown/tests.rs
expression: collection
---
version: 0.1.0
length: 1
value:
  - id: 0
    entity:
      uri: "https://foo.com/"
      createdAt: 1700006400
      updatedAt: []
      names:
        - Foo
      labels: []
      shared: false
      toRead: false
      isFeed: false
    edges: []
